eksperyment_mikroserwisy.rst

```diff
@@ -0,0 +1,139 @@
+Compose. Skonfigurowano je tak, aby jedną komendą można było uruchomić
+trzy serwisy podłączone do odpowiednich sieci, z odpowiednimi adresami
+zależności.
+
+Wnioski
+-------
+
+* Prosta funkcjonalność testowego systemu okazała się łatwa do
+  zrealizowania w architekturze mikrousług.
+
+* Frontend w modelu mikrousług można zrealizować następująco: każda
+  usługa, która chce mieć interfejs graficzny WWW ma go niezależnie od
+  innych usług (tzn. ma wbudowany serwer HTTP, który serwuje
+  strony). Jeśli jakaś usługa wewnątrz swojego frontendu chce załączyć
+  frontend innej mikrousługi, może zrobić to za pomocą iframe.
+
+* Gotowe mikrousługi można uruchamiać bez narzędzi
+  wirtualizacji/konteneryzacji. Można też uruchamiać je z użyciem
+  takich narzędzi. Użycie takich narzędzi jest dla usług
+  przezroczyste. To dobra cecha, ponieważ pozwala na zmianę środowiska
+  bez zmiany kodu usług.
+
+* Występują problemy z inicjalizacją trwałego stanu, co oznacza, że
+  inicjalizacja stanu jest szczególnym przypadkiem działania usługi,
+  który trzeba przewidzieć i obsłużyć (ten aspekt stosuje się do
+  wszelkich programów, które mają trwały stan). Są dwie główne
+  możliwości:
+
+  * Inicjalizacja jest przeprowadzana automatycznie, gdy aplikacja
+    wykryje konieczność inicjalizacji.
+  * Inicjalizacja jest przeprowadzana przez inny kod niż zasadnicza
+    aplikacja. Może to być inny program lub część aplikacji, która
+    uruchamiana jest przez przekazanie do aplikacji odpowiedniego
+    argumentu.
+
+* Korzystanie z zarządcy zależności nie wygląda na dobry pomysł. Każda
+  usługa powinna przyjmować w swoich argumentach lokalizację innych
+  aplikacji, od których zależy i nie korzystać z zarządcy zależności.
+
+* Dzięki konteneryzacji nie ma problemu z kolidującymi wersjami
+  oprogramowania (np. A wymaga biblioteki foobar-3.3.6, B wymaga
+  biblioteki foobar-3.4.1), nie ma problemu z kolizjami portów
+  (wszystkie serwery HTTP, których jest n, można uruchomić na porcie
+  80).
+
+* Do zarządzania zależnościami pomiędzy usługami dobrze sprawdza się
+  DNS serwowany przez Dockera wewnątrz sieci wirtualnych,
+  np. uruchamiamy mikrousługę A i nazywamy ją "ms_a". Następnie,
+  uruchamiając mikrousługę B, która korzysta z A, podajemy jej adres
+  A: http://ms_a/api/. Nie musimy wiedzieć jaki adres ma
+  usługa A. Dzięki temu adres może być przydzielany dynamicznie.
+
+* Docker-compose znacznie ułatwia uruchamianie projektu poprzez
+  automatyzację uruchamiania wielu powiązanych usług.
+
+* Być może Docker nie powinien obsługiwać stanowych
+  usług. Stwierdzenie nie jest wynikiem eksperymentu, lecz artykułów
+  wskazujących zagrożenia związane z uruchamianiem stanowych usług w
+  kontenerach (por. `artykuł 1`_, `artykuł 2`_). Cytaty:
+
+  * "Docker MUST NOT run any databases in production, EVER."
+  * "Services which store data cannot be dockerized. Docker is
+    designed to NOT store data. Don’t go against it, it’s a recipe for
+    disaster."
```

Comment:
Pragnę podkreślić, że rozwiązanie kwestii aplikacji stanowych jest istotne i krytyczne. Dostawcy mogą chcieć wykorzystać nie tylko kilka rodzajów baz danych (PostgreSQL / Oracle / MariaDB), ale także korzystać z wyszukiwarek (Elasticsearch), korzystać z współdzielonego cache (Redis itd.), czy w ogóle przechowywać jakieś pliki. Pliki są w tego rodzaju bezstanowych środowiskach przechowywane w storage obiektowym i przekazywane np. z pomocą  API S3 / Swift. 

Warto dostrzec, że platforma konterów Heroku działa na platformie Amazon AWS. Można korzystać z gotowych usług stanowych (co oferować mogłaby platforma MC), ale dzięki temu, że działa to w AWS zawsze można dostawić usługę w danej strefie na wirtualnej maszynie i łączyć środowiska.